### ISSUE TITLE
Global constants without `global::`

### ISSUE THREAD

**Issue by RazrFalcon** [rhaiscript/rhai#481]
Using `global::` to access global constants inside functions looks like a hack. Is there a way to have truly global constants?

**Comment by schungx** [comment]
Not really. Rhai is structured specifically such that all functions are isolated from each other.  This makes them easy to mix 'n match.

In order to avoid having to search through a scope chain, Rhai functions have no access to global constants (which essentially must form a scope of itself).

The engine currently _hacks_ it by copying the values of constants into a separate module called `global`.

**Comment by RazrFalcon** [comment]
Can't you simply perform constants propagation?

I would prefer constants to be resolved during parsing, so when I'm calling and individual function from the script - it would still work.

**Comment by schungx** [comment]
Not into functions currently... functions are completely isolated from each other, even from global constants.

If it is just for optimization purposes, I suppose something can be done to make those constants available to the optimizer...

**Comment by RazrFalcon** [comment]
No, it's not about performance, but about usability. I just want some sort of truly global constants.

**Comment by schungx** [comment]
Maybe I can add a `static` statement which declares global constants that will always be propagated to the entire script during parsing?

Will that work?

**Comment by RazrFalcon** [comment]
Isn't this what `const` should do? What would be the difference between `const` and `static` then?

**Comment by schungx** [comment]
`const` is essentially the same as `let`...  it only won't allow you to change the variable's value later on.

`static` in this case would be much more restrictive.  For example, it probably doesn't make sense for `static` statements within a localized block, or within a function; only at top-most scope.

**Comment by RazrFalcon** [comment]
I see. In Rust terms, I need `const`, not `static`. Or `#define VALUE 45;` in C. 

For some reason, I thought that `const` in Rust is a global value. Maybe because you can access it via `global::`. But I guess this is just  a constant variable.

**Comment by schungx** [comment]
Actually, I can see it now.  `Engine::eval_with_scope`, `Engine::run_with_scope` and `Engine::compile_with_scope()` are supposed to do sorta what you want - any constants inside the scope passed in will propagate throughout the script, including functions.

However, there is a bug and functions are skipped.

I've fixed it, so in the next minor version you can use `eval_with_scope` with a custom scope containing your constant.

Unfortunately you still cannot do:

```rust
const VALUE = 42;

fn foo() { VALUE }

print(foo());   // hopefully this will print "42"
```

**Comment by RazrFalcon** [comment]
All my constants are defined in the script itself. I cannot set them from Rust.

**Comment by schungx** [comment]
Can you search for all lines starting with `const` and then extract them into a script, then compile it, and evaluate it into a scope.

Iterate the variables inside that scope and add them as constants into a new scope.  This gives you the scope with the constants.

Then you can just do `eval_with_scope` on the original script to propagate those constants into functions.

**Comment by RazrFalcon** [comment]
I have no plans in writing a Rhai preprocessor 😄

**Comment by schungx** [comment]
It probably won't be too hard.  Pre-processing on text scripts is highly common.

It probably is going to be 20-30 lines of Rust using `regex`.

**Comment by schungx** [comment]
I can probably even do it here:

```rust
let script = " ... ";      // The original script

let constants = script.lines().filter(|s| s.starts_with("const ")).collect::<Vec<_>>().join("\n");

let mut scope = Scope::new();

engine.eval_with_scope(&mut scope, constants);

// ... now we have all the global constants inside the scope

engine.eval_with_scope(&mut scope, script);      // you'll need a newer version of Rhai that fixes the constants propagation bug
```

**Comment by schungx** [comment]
If you'd like to pull from https://github.com/schungx/rhai, you can try out a new API that I added for this purpose.

```rust
let ast = engine.compile(" ... script ...")?;

// Create a Scope with all top-level constants
let scope: Scope = ast.iter_literal_variables(true, false).collect();

// Re-optimize the AST to propagate constants into functions
let ast = engine.optimize_ast(&scope, ast, OptimizationLevel::Simple);

// ... now the AST is optimized
```

**Comment by schungx** [comment]
Closing this for now unless there are further issues.

**Comment by profan** [comment]
> If you'd like to pull from https://github.com/schungx/rhai, you can try out a new API that I added for this purpose.
> 
> ```rust
> let ast = engine.compile(" ... script ...")?;
> 
> // Create a Scope with all top-level constants
> let scope: Scope = ast.iter_literal_variables(true, false).collect();
> 
> // Re-optimize the AST to propagate constants into functions
> let ast = engine.optimize_ast(&scope, ast, OptimizationLevel::Simple);
> 
> // ... now the AST is optimized
> ```

Testing this out right now using the specific setup in this comment and it doesn't seem like constants have been propagated into my function, or at least my script errors out on not finding the constant that should habeen propagated, is this still the right way to do it?

I'll offer more context if necessary :)

Script compilation is set up like this:
![Code_zdbyt0Rpmk](https://user-images.githubusercontent.com/2816340/230686800-204bff56-b463-4766-beed-016045f9305b.png)

Script evaluation later happens like this:
![image](https://user-images.githubusercontent.com/2816340/230686875-48ea8ec9-1c84-4307-9468-75988f331500.png)

Anything obviously wrong or want me to provide more context or try to put together a repro case?

**Comment by schungx** [comment]
What does your script looks like?  Can you check if the constant variables are successfully collected into the `Scope`?  You can simply inspect `scope.len()`.

**Comment by profan** [comment]
> What does your script looks like? Can you check if the constant variables are successfully collected into the `Scope`? You can simply inspect `scope.len()`.

Yeah, I checked the contents of the scope used for the call to optimize_ast and all my constants seemed to be there.

**Comment by schungx** [comment]
That narrows it. I'll check it out.

**Comment by schungx** [comment]
OK, it was a bug.  After constants propagation, I forgot to remove the variables sharing operation, and Rhai is still trying to share the variable with that name.

**Comment by schungx** [comment]
@profan   I have uploaded a fix.  You can pull from this repo to test it out.  Thanks for catching this.

**Comment by profan** [comment]
> @profan I have uploaded a fix. You can pull from this repo to test it out. Thanks for catching this.

Thank you for the fix! I'll set my cargo dep to master with the fix here and give it a try later today :)

**Comment by profan** [comment]
> @profan I have uploaded a fix. You can pull from this repo to test it out. Thanks for catching this.

Seems like the exact same issue still persists for me, I updated the dependency of rhai on my repository to main here, made sure to clean and update my cargo deps, the lockfile seems to reflect the right version, so I think something is still off somewhere 🤔 

I updated my repository to directly reflect the issue here: https://github.com/profan/some-macroquad-things/commit/1a8561a659114b63b0273e18fe3965fbaa9d5523

If you want to use that as a case/see if there's something strange that I'm doing specifically

**Comment by schungx** [comment]
It would help to have a minimally reproducible sample that still contains the error.

I see that it is not an easy task to compile your repo just to test out the scripting problem...

**Comment by schungx** [comment]
Turns out this has been a really subtle bug, which I now caught, thanks to your report.

You can pull from this repo again to see if it finally solves your problem.

**Comment by profan** [comment]
> Turns out this has been a really subtle bug, which I now caught, thanks to your report.
> 
> You can pull from this repo again to see if it finally solves your problem.

Thanks for your patience! So that has definitely solved the immediate constants issue, but now I have a fun one which I assume might actually be designed behaviour, so I have this pile of constants at the top of the file, but presumably these are all evaluated when the AST is built, so you can't have constants that call into rust functions for their initial state I assume?

ie. something like this (something like HEAD_ROTATION here):
![image](https://user-images.githubusercontent.com/2816340/230892414-25546cb7-9129-4b73-a376-ef8c1c7e5337.png)

**Comment by profan** [comment]
> You're correct. Constants propagation occurs during _parsing_, which won't run your AST.
> 
> Unless you use `OptimizationLevel::Full` that is... If you use full optimizations, functions with all-constant arguments are eagerly-called. In this case, your `quat_identity` will end up being called at optimization time.

Cool, well this matter is definitely solved for me as it is then, appreciate all the help!

**Comment by schungx** [comment]
> Cool, well this matter is definitely solved for me as it is then, appreciate all the help!

No prob!  Would be interested to know if full optimizations work out for you.  I really haven't seen anybody using Full so I'm really curious how it goes.

**Comment by profan** [comment]
> > Cool, well this matter is definitely solved for me as it is then, appreciate all the help!
> 
> No prob! Would be interested to know if full optimizations work out for you. I really haven't seen anybody using Full so I'm really curious how it goes.

I think the way optimization level full seems to work, I probably won't be using it because I have a lot of functions that depend on the "current" state like get_current_time etc (which are very much not pure, but rhai doesn't know that), it might make sense to have an optimization level where it evaluates constants at compilation time, but doesn't go any further to assuming all functions with constant arguments are pure, or maybe the user could be offered some way of specifying that a function is a "pure" function in the function registration API?

At least that's what would make me actually use full optimizations, right now it looks much too scary to try to use it :)

**Comment by schungx** [comment]
Well, what you can do is to have two separate engines.  One only for compiling and contains all the simple functions such as `vec3` and `vec3 + vec3`.  The ones you know definitely have no side-effects.

Then you can use it to full-optimize your AST.

**Comment by profan** [comment]
> Well, what you can do is to have two separate engines. One only for compiling and contains all the simple functions such as `vec3` and `vec3 + vec3`. The ones you know definitely have no side-effects.
> 
> Then you can use it to full-optimize your AST.

I mean, I definitely _could_, but I reckon I won't, because it sounds like more effort than gain 😅

**Comment by schungx** [comment]
> If I at registration time could say "this is a function without side effects/this function only depends on its parameters, you're free to evaluate this at compile time as long as it has constant arguments", that would make a lot more sense overall to me as a user.

We have two slightly different definitions of "pure" in Rhai:

1) The function does not mutate its first `&mut` parameter (or has no such parameter), but _may_ cause side effects (e.g. via interior mutability)

2) The function does not mutate its first `&mut` parameter (or has no such parameter), and does not cause any addition side effects

For optimization purpose, we'd need the tighter No.2 definition, while for script execution we only need No.1 (just to make sure a function is not mutating a constant).

Rhai does not have No.2 definition pure right now, so it'd have to be added and an API to turn it on/off.

**Comment by schungx** [comment]
> I mean, I definitely _could_, but I reckon I won't, because it sounds like more effort than gain keeping these things in sync 😅

Not necessarily...  it would have substantial implications on performance, especially when some of your functions are simple ones such as constructors (e.g. `vec3`) and simple ops (e.g. `+`).  If you don't optimize them, every run of your script will be calling those constructors to construct constant objects that you could have constructed once during optimization.

That depends on how _many_ of those you have in your script and how frequently you intend to call your script.

**Comment by profan** [comment]
> > I mean, I definitely _could_, but I reckon I won't, because it sounds like more effort than gain keeping these things in sync 😅
> 
> Not necessarily... it would have substantial implications on performance, especially when some of your functions are simple ones such as constructors (e.g. `vec3` or matrix constructors such as `translate`, `rotate`, `scale`) and simple ops (e.g. `+`). If you don't optimize them, every run of your script will be calling those constructors to construct constant objects that you could have constructed once during optimization.
> 
> That depends on how _many_ of those you have in your script and how frequently you intend to call your script.
> 
> For example, as in your script:
> 
> ```rust
> const POSITION = rotate(PI / 2) * scale(2) * vec3(1, 2, 3) + vec3(4, 5, 6);
> const NORMAL = POSITION.normal;
> const PROJECTION = POSITION * NORMAL;
> ```
> 
> If you don't optimize them, every run of your script will go through a bunch of calculations just to populate `POSITION`, `NORMAL` and `PROJECTION`. If you full-optimize them, all three will be replaced by constant `vec3` objects (or constant values).

Yeah, but I also don't want to subtly break my scripts if full optimizations does something unexpected, so ergonomics > performance, especially when it's not currently a problem :eyes:

I may look into it later if it actually does become a problem, but I'm more likely to just move more stuff into rust proper

**Comment by profan** [comment]
> > If I at registration time could say "this is a function without side effects/this function only depends on its parameters, you're free to evaluate this at compile time as long as it has constant arguments", that would make a lot more sense overall to me as a user.
> 
> We have two slightly different definitions of "pure" in Rhai:
> 
>     1. The function does not mutate its first `&mut` parameter (or has no such parameter), but _may_ cause side effects (e.g. via interior mutability)
> 
>     2. The function does not mutate its first `&mut` parameter (or has no such parameter), and does not cause any addition side effects
> 
> 
> For optimization purpose, we'd need the tighter No.2 definition, while for script execution we only need No.1 (just to make sure a function is not mutating a constant).
> 
> Rhai does not have No.2 definition pure right now, so it'd have to be added and an API to turn it on/off.

I see, yeah I was looking around the rhai docs and did see the pure_fn annotations as you mention with number 1, so was wondering what the overlap there was, so this makes sense, it is definitely a bunch of extra complexity to have something like the second form of purity have to live in rhai as well 🤔
